stl/inc/future

```diff
@@ -752,12 +749,34 @@ public:
     }
 
     _State_manager& operator=(const _State_manager& _Other) noexcept {
-        _Copy_from(_Other);
+        // copy stored associated asynchronous state object from _Other
+        if (this != _STD addressof(_Other)) {
+            if (_Assoc_state) {
+                _Assoc_state->_Release();
+            }
+
+            if (_Other._Assoc_state) { // do the copy
+                _Other._Assoc_state->_Retain();
+                _Assoc_state   = _Other._Assoc_state;
+                _Get_only_once = _Other._Get_only_once;
+            } else {
+                _Assoc_state = nullptr;
+            }
+        }

```

Comment:
I _hate_ self-checks in assignment operators, they pessimize the common case. We can get rid of this one with the "increment the reference count before decrementing" pattern:
```suggestion
        if (_Other._Assoc_state) {
            _Other._Assoc_state->_Retain();
        }

        if (_Assoc_state) {
            _Assoc_state->_Release();
        }

        _Assoc_state   = _Other._Assoc_state;
        _Get_only_once = _Other._Get_only_once;
```

Reply:
This may have negative impact of performance on self-copy-assignment, though...

Reply:
I'm happy to pessimize a rare corner case to reduce total codesize and speedup the common case.